### ISSUE TITLE
textDocument/linkedEditingRange failed when opening files or moving cursor

### ISSUE THREAD

**Issue by mfru** [BetterThanTomorrow/calva#1374]
Observed behavior: Calva throwing an error, but still working.

Expected behavior: Calva working without throwing errors or printing a more helpful error message.

When opening my Leiningen project in VS Code I get

```
Request textDocument/linkedEditingRange failed.
  Message: No symbol found
  Code: -32602 
```

every time I open a `.clj` file or move the cursor inside a `.clj` file.

Otherwise it is working fine.

Stacktrace in VS Code DevTools:
```
 ERR No symbol found: Error: No symbol found
	at /home/mfru/.vscode-server-insiders/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1012761
	at /home/mfru/.vscode-server-insiders/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1013055
	at Immediate.<anonymous> (/home/mfru/.vscode-server-insiders/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1013420)
	at processImmediate (internal/timers.js:461:21)
```

The last time I remember everything being ok (last time I opened the project) was about a week ago, so maybe a VS Code or Calva update introduced the error?

```
Calva version: 2.0.223

VS Code Insider version: 
Version: 1.62.0-insider (user setup)
Commit: 4bbec283c36a51cf80f9b77c7a81c140a76a363b
Date: 2021-11-02T08:19:40.220Z
Electron: 13.5.1
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19043

Working inside WSL2 with Ubuntu 20.04.3 LTS on Windows 10 x86_64

Leiningen 2.9.3 on Java 11.0.11 OpenJDK 64-Bit Server VM
```

**Comment by mfru** [comment]
I rolled back down to version `2.0.216` of Calva which makes the error go away (everything above doesn't), don't know how that makes sense given the working state of about a week ago, which should have been an already affected version.

It might be related to the issue described in #1373

**Comment by bpringe** [comment]
Thanks for reporting this. It does seem to be related to that issue, in particular to the updating of the LSP client library. `textDocument/linkedEditingRange` is a LSP request method. 

@ericdallo, any idea what could be going wrong here?

**Comment by ericdallo** [comment]
@bpringe yes, we added on last release the live rename via `textDocument/linkedEditingRange` but I can't understand why vscode would call it without manual user action...

**Comment by ericdallo** [comment]
I'd like to understand why vscode calls this method without user interaction, if that makes sense for some reason, we could make clojure-lsp not return error but a null value when not in a symbol

**Comment by mfru** [comment]
Are you able to reproduce the error as is? If not I'll try to find a minimal reproduction setup.

Until then steps to reproduce _should_ be

1. Have VSCode `1.62.0-insider` with Calva `2.0.217 - 2.0.223` installed
2. Open a Leiningen project (maybe a single .clj file suffices as well)

**Comment by ericdallo** [comment]
I can't repro that with a lein project and calva 2.0.222, but I'm using vscode `1.60.2` but I doubt it could be related

**Comment by mfru** [comment]
Tried to repro with the portable version of VSCode `1.62.0` with Calva `2.0.223` but it throws no error.

Will have to be an individual issue on my existing setup then.

I'll let you know if I find anything out.

**Comment by bpringe** [comment]
To provide another data point, I tried to reproduce as well with Calva `2.0.223` and VS Code Insiders `1.63.0` as well as VS Code `1.62.0` and could not.

**Comment by gAdrev** [comment]
I am also getting this with:

- VS Code `1.62.1`
- All extensions disabled except for Calva
- Calva `2.0.223`
- OS: Linux, `deb` vscode repo install (Ubuntu 20.04)
- Open a folder consisting of a single `deps.edn` file with contents: `{}`

Just opening the `deps.edn` file makes the toast appear.

The extension log reads this for each toast:

```
[2021-11-10 01:12:04.501] [exthost] [error] [betterthantomorrow.calva] provider FAILED
[2021-11-10 01:12:04.502] [exthost] [error] Error: No symbol found
	at /home/javier/.vscode/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1012761
	at /home/javier/.vscode/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1013055
	at Immediate.<anonymous> (/home/javier/.vscode/extensions/betterthantomorrow.calva-2.0.223/out/extension.js:2:1013420)
	at processImmediate (internal/timers.js:461:21)
```

As @mfru said, downgrading to `2.0.216` makes the error go away. Anything starting from `2.0.217` or newer up to current `2.0.223` produces the error.

**Comment by gAdrev** [comment]
Also, testing with VSCode Insiders `1.63.0` (deb packaged) produces the same error with the newer versions of Calva.

**Comment by ericdallo** [comment]
thanks for the feedback, this still stands for me: https://github.com/BetterThanTomorrow/calva/issues/1374#issuecomment-961185038
The worst case calva could disable that linkedEditingRange feature manually

**Comment by bpringe** [comment]
[This search](https://github.com/microsoft/vscode-languageserver-node/search?q=linkedEditingRange) might be a good place to start. This only started happening after `vscode-languageclient` was updated in [this PR](https://github.com/BetterThanTomorrow/calva/pull/1336).

I only looked briefly in that `vsode-languageserver-node` repo (which includes the client library), and I didn't see anything pointing to it being called by the client, but further inspection might yield useful info.

One other thing that I suppose could be happening is that something Calva or VS Code are calling causes that request to be made when the 7.0.0+ version of the client is available. I don't know if that makes sense though.

**Comment by bpringe** [comment]
Accidentally closed 😄

**Comment by ericdallo** [comment]
> One other thing that I suppose could be happening is that something Calva or VS Code are calling causes that request to be made when the 7.0.0+ version of the client is available. I don't know if that makes sense though.

I really think this is true, calva started calling after the version was upgraded + clojure-lsp added support to linkedEdittingRange on latest release

**Comment by bpringe** [comment]
Adding some info from Eric in Slack:

> I think there is a quickfix here of force disabling linkedEdittingRange on calva side
> Maybe forcing to not send  that capability when initializing server

This probably wouldn't take long to fix. I'll prioritize it, but it might be a bit before I get to it. A PR is welcome if anyone feels the desire to do it before me.

**Comment by bpringe** [comment]
From Rick Jensen on Slack:

> Interestingly enough, my machine crashed and after rebooting the problem seems to have gone away. Not sure if something updated behind the scenes or it is due to a race condition, but I’m no longer experiencing the buggy behavior I was before.

I'm not sure why it would stop happening like that, but I wanted to add the additional info here.

**Comment by bpringe** [comment]
I've re-enabled linkedEditingRange in #2703, but I'd like others to test the vsix since I wasn't able to reproduce the issue in the past. Let me know if the issue still happens!

https://output.circle-artifacts.com/output/job/cffcf6e4-309a-475d-b170-3d09680a1363/artifacts/0/tmp/artifacts/calva-2.0.484-1374-enable-linkedEditingRange-20386c91.vsix

CC @ericdallo @PEZ @mfru @gAdrev

**Comment by bpringe** [comment]
Here are instructions for testing a vsix, if needed: https://github.com/BetterThanTomorrow/calva/wiki/Testing-VSIX-Packages

**Comment by paulbutcher** [comment]
I just tried the vsix. With Linked Editing enabled (everything works exactly as expected with it disabled, which it is by default):

<img width="732" alt="Screenshot 2025-01-13 at 09 36 47" src="https://github.com/user-attachments/assets/ec02603f-cb1c-4881-8c68-f6d8fafc5f84" />

I get this as soon as I open a Clojure file:

```
[Error - 09:37:28] Request textDocument/linkedEditingRange failed.
  Message: No symbol found
  Code: -32602 
[Error - 09:37:28] Request textDocument/linkedEditingRange failed.
  Message: No symbol found
  Code: -32602 
```

That being said, it does kinda work (I'm not sure why sometimes the above failure happens and sometimes it doesn't: it seems to vary depending on the exact state of the buffer) and I see exactly the same behaviour as I do in Zed, as discussed in https://github.com/clojure-lsp/clojure-lsp/discussions/1936

https://github.com/user-attachments/assets/6dc41733-573f-4e44-9f58-7205510d0c31

**Comment by paulbutcher** [comment]
OK, the explanation for the errors is quite simple: [`clojure-lsp.feature.linked-editing-range/ranges`](https://github.com/clojure-lsp/clojure-lsp/blob/master/lib/src/clojure_lsp/feature/linked_editing_range.clj#L17) always returns an error if the cursor isn't on a symbol (or some other condition means it can't find a symbol under the cursor). 

It should, I believe, return `null` in these circumstances:

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_linkedEditingRange

**Comment by paulbutcher** [comment]
It's not as simple, sadly, as just returning success in those situations: there's definitely still something else wrong.

Modifying simple symbols works (as shown in the video above. But modifying aliases fails. Here's what happens in VSCode:

https://github.com/user-attachments/assets/85a26860-f709-49e8-be9f-6102e6d322eb

Editing the alias changes the *entire* referenced symbol (not just the namespace). And then things, unsurprisingly break from that point onwards.

In Zed, the behaviour is slightly different (although not correct):

https://github.com/user-attachments/assets/b57a47ad-33b0-4f0a-9cc1-7e941a5d0335

Only the namespace portion of the referenced symbol is modified, but there seems to be some kind of timing sensitivity: the first keystroke always works, but subsequent keystrokes might or might not (typing slowly seems to be more reliable). In the video above the first two keystrokes worked but the third didn't (my experience is different each time).

All things considered, given the limited utility of linked edits for Clojure, I'm increasingly coming to the conclusion that the best course of action is to simply remove them from clojure-lsp. What do you think?

**Comment by bpringe** [comment]
> All things considered, given the limited utility of linked edits for Clojure, I'm increasingly coming to the conclusion that the best course of action is to simply remove them from clojure-lsp. What do you think?

@ericdallo ^

**Comment by ericdallo** [comment]
After that comment @paulbutcher helped with pra that improved linkedEdit, so not sure if it's that big deal that requires remove the feature

**Comment by paulbutcher** [comment]
For completeness, it's definitely much better:

* Linked edits now only apply to namespace aliases (so the "blast radius" of any issues is much smaller).
* We've fixed the issue of changing the entire referenced symbol instead of just the namespace alias portion.
* We've improved (although not entirely fixed) the timing sensitivity.

However:

* There are still some timing sensitivities. For some reason these only show up in VSCode and not in Zed (I don't know about other editors). We don't yet know where these come from.
* There are still issues with aliases that include hyphen (`-`) characters in Zed. This is a [bug in Zed](https://github.com/zed-industries/zed/issues/23286), not clojure-lsp, but still a problem that significantly affects the utility of the feature.
